### ISSUE TITLE
Workflow Tracking

### ISSUE THREAD

**Issue by ccpony** [danielgerlag/workflow-core#18]
I started an issue, but then I didn't follow up.  Let me try again...

What is the best way to track long-running workflows, in terms of the state of the individual steps, etc.?  At one point we had considered a separate table for persisting each step.  I think that this would be a good idea since user-definable data could be persisted on a per-step basis.  Ultimately, it would be nice to provide some kind of front end to all this (such as GoJS).

Daniel - do you have any plans for this? 

Thanks again.

**Comment by danielgerlag** [comment]
Hi Jeff,

Sorry for the delayed response... I have recently changed the way the relational DB persistence providers store their data. (SQL Server, Postgres & Sqlite), so that more of the details are split out into separate tables.
The "ExecutionPointer" table is what you are looking for, this will hold info on which steps have been executed and what the next step to execute is.  Also, if you look at the WorkflowCore.Users project under extensions, you can see an example of how I have added additional data on a per step basis, that is not part of the actual core library.  Let me know if this works for you?

I was planning on building an admin UI in Angular 2, but this will probably be a separate repo, also still need to refine the API project first.

**Comment by ccpony** [comment]
Hello Daniel.

The project I’m working on uses WorkflowCore on the back-end with Net Core/ASP.Net Web Api and Sql Server.  The front end is Angular 2 and Google Material Design and many adjunct libraries.  It’s an enterprise application called Proximity that manages the procurement and distribution of things directly to employees of multi-location, distributed organizations (i.e., uniforms, safety equipment, replacement parts, tools, etc., etc.).  The software will negate the need for organizations to manage their own central (expensive/cumbersome/prone-to-failure) inventory systems.  Workflows are central to this application since there are numerous complex approval processes, order handling processes, catalog/program management processes, etc.  It will eventually be deployed on Azure.

I’m pretty proficient on Angular 2/Material UI.  Please let me know if I can help with the Admin UI.  That would definitely be the area that I could be of most use.

Finally, on the topic of workflow tracking, thanks for your response.  I’ll check out the code that you’ve mentioned and let you know if that will work.

If you have any ideas regarding a browser-based UI for in-flight workflows (i.e., diagram that shows the overall workflow structure and the current state of each step (perhaps color coded).  Includes, perhaps, step completion information such as date/time, exceptions and even user-defined information such as the name of the player who initiated the event/trigger).  Please let know if this is something that is worth pursuing.  I would be very interested in providing end-users with a visual representation of any of the many long-running workflows that will be active at any given time.

Thanks again for your incredibly work on this.

Jeff

**Comment by danielgerlag** [comment]
Hi Jeff,

Were you able to solve your problem with the new schema for the SQL provider?

Also, I'd like to take you up on the offer to build an admin UI, that would be awesome.  I was thinking, let's keep it in this repo but put it under the "extensions" directory.  Feel free to modify and extend the web API library as needed to support this.

**Comment by ccpony** [comment]
Hello Daniel.  Please advise how I can discuss a business opportunity with you privately.  Thanks.

Jeff

**Comment by danielgerlag** [comment]
daniel@gerlag.ca
+1 604 290 3853
